src/coreclr/vm/profilingenumerators.cpp

```diff
@@ -552,9 +552,11 @@ HRESULT ProfilerThreadEnum::Init()
     }
     CONTRACTL_END;
 
+    // If EnumThreads is called from a profiler callback where the runtime is already suspended,
+    // don't recursively acquire/release the ThreadStore Lock.
     // If a profiler has requested that the runtime suspend to do stack snapshots, it
     // will be holding the ThreadStore lock already
-    ThreadStoreLockHolder tsLock(!g_profControlBlock.fProfilerRequestedRuntimeSuspend);
+    ThreadStoreLockHolder tsLock(!GCHeapUtilities::IsGCInProgress() && !g_profControlBlock.fProfilerRequestedRuntimeSuspend);
```

Comment:
If the intent is to not acquire thread store lock second time, then this should check if the current thread is already holding the thread store lock. 

GC could be started by other thread and GC may be changing the thread list (for example starting or background threads). 
If current thread is not holding a thread store lock, it should acquire it before iterating through thread list.

Reply:
Thanks for pointing that out! I was mainly looking to not acquire the thread store lock a second time. Will change to `ThreadStore::HoldingThreadStore()`